### PR TITLE
Update project trackers on redmine server

### LINKED ISSUE #318: Project.addTrackers does not update redmine database
Trackers can be added to a project object, however after calling projectManager.update the new trackers have not been added to Redmine.

It seems that java api tries to add tracker objects, but Redmine api requires an array of tracker ids.
Some code to test this, latest assert fails when checking if Redmine has been updated
```
    @Test
    public void testUpdateTrackers() throws RedmineException {
        Project projectToCreate = generateRandomProject();
        String createdProjectKey = null;
        try {
            Project createdProject = projectManager.createProject(projectToCreate);
            createdProjectKey = createdProject.getIdentifier();
            Collection<Tracker> trackers=createdProject.getTrackers();
            assertEquals("assume redmine is configured with 1-Bug as default tracker", 1, trackers.size());
            
            Collection<Tracker> newTrackers=new ArrayList<Tracker>();
            newTrackers.add(TrackerFactory.create(2,"Feature"));
            newTrackers.add(TrackerFactory.create(3,"Support"));
            createdProject.addTrackers(newTrackers);
            assertEquals("trackers should be added to bean", 3, createdProject.getTrackers().size());
            
            projectManager.update(createdProject);
            createdProject=projectManager.getProjectByKey(createdProjectKey);
            assertEquals("trackers should be added to redmine", 3, createdProject.getTrackers().size());
        } finally {
            if (createdProjectKey != null) {
                projectManager.deleteProject(createdProjectKey);
            }
        }
    }
```

### PR DESCRIPTION
**Pull request by javiertuya**
Fixes #318.

Cause: Trackers are not updated in the server because RedmineJSONBuilder.writeProject sends a collection of Tracker objects to redmine server. However the api requires an array of tracker ids.

Solution: Writing trackers uses JsonOutput.addScalarArray wich replaces JsonOutput.addArrayIfNotNull.

But this still does not solve the problem, as in the case of a new project, an empty array of trackers would be sent to the server. Therefore any default tracker configured in the server would be overridden, and the project be created without trackers.

To avoid this, now the Project constructor does not initialize the storage for TRACKERS (initializations are made in project’s getters and setters that manage trackers). An additional method to clear all trackers has been added.

### CODE REVIEW

File: src/main/java/com/taskadapter/redmineapi/ProjectManager.java [hunk 1]
```diff
@@ -64,7 +64,14 @@
      * 	Project createdProject = mgr.createProject(project);
      * }
      * </pre>
-     *
+     * 
+     * <p>
+     * Note: if {@code project} trackers have not been set with {@link Project#addTrackers}
+     * or cleared with {@link Project#clearTrackers},
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/ProjectManager.java]: I would reword this for clarify to indicate that "or cleared" actually means "or if they have been cleared".

File: src/test/java/com/taskadapter/redmineapi/ProjectIntegrationIT.java [hunk 2]
```diff
@@ -209,12 +211,71 @@ public void tryUpdateProjectWithLongHomepage() throws RedmineException {
...
+            createdProject.addTrackers(trackers);
+            projectManager.update(createdProject);
+            createdProject=projectManager.getProjectByKey(createdProjectKey);
+            assertEquals(1, createdProject.getTrackers().size());
+
+            //add more than one tracker, it does not replace previous trackers
+            trackers=new HashSet<Tracker>(Arrays.asList(availableTrackers.get(1), availableTrackers.get(2)));
+            createdProject.addTrackers(trackers);
+            projectManager.update(createdProject);
+            createdProject=projectManager.getProjectByKey(createdProjectKey);
```
Comment (alexeyOnGitHub) [inline src/test/java/com/taskadapter/redmineapi/ProjectIntegrationIT.java]: let's not reuse vars once they are defined. use separate variables for readability instead

File: src/main/java/com/taskadapter/redmineapi/bean/Project.java [hunk 3]
```diff
@@ -99,13 +98,29 @@ public void setName(String name) {
...
     }
 
+    /**
+     * Adds the specified trackers to this project.
+     * If this project is created or updated on the redmine server, 
+     * each tracker id must be a valid tracker on the server.
+     */
     public void addTrackers(Collection<Tracker> trackers) {
+    	if (!storage.isPropertySet(TRACKERS)) //checks because trackers storage is not created for new projects
+    		this.clearTrackers();
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/bean/Project.java]: if property is not set, then clear something - this is confusing. 
why clear if it is not set?

correspondingly, similar comment about the change in getTrackers() method. 
can we keep existing ```storage.set(TRACKERS, new HashSet<>());``` (I see it was removed here)   
and use that one instead?

File: src/main/java/com/taskadapter/redmineapi/ProjectManager.java (hunk 1 above)
Comment (javiertuya) [reply src/main/java/com/taskadapter/redmineapi/ProjectManager.java]: Done. Now the or changes to and:

> if trackers have not been set with addTrackers and if they have been cleared with clearTrackers, the created project will get the server default trackers (if any). Otherwise, the project trackers will override the server default settings.

The comment is moved to Project.create() after rebase.

File: src/test/java/com/taskadapter/redmineapi/ProjectIntegrationIT.java (hunk 2 above)
Comment (javiertuya) [reply src/test/java/com/taskadapter/redmineapi/ProjectIntegrationIT.java]: Done, tests updated to fluent style

File: src/main/java/com/taskadapter/redmineapi/bean/Project.java (hunk 3 above)
Comment (alexeyOnGitHub) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: what about this change?

Comment (javiertuya) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: Sorry, I forgot commit the comment.
addTrackers(): I have replaced this.clearTrackers() by storage.set(TRACKERS, new HashSet<>())

getTrackers(): new HashSet<>() is returned in order to avoid modifying the bean (as it is a getter). If we set the storage to an empty set of trackers and next operation is creating the project, it will be created without default trackers.

Comment (alexeyOnGitHub) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: I see the idea, but I am worried about getTrackers() method returning fake data. 
users can attempt to call
.getTrackers() and then add an item to the collection, and the behavior would depend on whether or not there were trackers before: it would return either a new **modifiable** hashset, or an **unmodifiable** set if there were some trackers stored already.
one option would be to return an unmodifiable empty set

Comment (javiertuya) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: I agree. Then, the change should be
```
    public Collection<Tracker> getTrackers() {
    	if (!storage.isPropertySet(TRACKERS)) //checks because trackers storage is not created for new projects
    		return Collections.unmodifiableCollection(new HashSet<Tracker>());
        return Collections.unmodifiableCollection(storage.get(TRACKERS));
    }
```
Is this right?

Comment (alexeyOnGitHub) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: this would work

Comment (javiertuya) [reply src/main/java/com/taskadapter/redmineapi/bean/Project.java]: Done. Thank you for your review.
